src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -76888,5 +76900,1347 @@ A M()
+                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "new C(z = 1)").WithArguments("C", "Deconstruct").WithLocation(7, 34),
+                // (7,34): error CS8129: No suitable 'Deconstruct' instance or extension method was found for type 'C', with 2 out parameters and a void return type.
+                //         (object? x, object? y) = new C(z = 1);
+                Diagnostic(ErrorCode.ERR_MissingDeconstruct, "new C(z = 1)").WithArguments("C", "2").WithLocation(7, 34));
+        }
+
+        [Fact]
+        public void Deconstruction_TooFewDeconstructArguments()
+        {
+            var source =
+@"class C
+{
+    internal void Deconstruct(out object x, out object y, out object z) => throw null;
+}
+class Program
+{
+    static void F()
+    {
+        (object? x, object? y) = new C();
+        x.ToString();
+        y.ToString();
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (9,34): error CS7036: There is no argument given that corresponds to the required formal parameter 'z' of 'C.Deconstruct(out object, out object, out object)'
+                //         (object? x, object? y) = new C();
+                Diagnostic(ErrorCode.ERR_NoCorrespondingArgument, "new C()").WithArguments("z", "C.Deconstruct(out object, out object, out object)").WithLocation(9, 34),
+                // (9,34): error CS8129: No suitable 'Deconstruct' instance or extension method was found for type 'C', with 2 out parameters and a void return type.
+                //         (object? x, object? y) = new C();
+                Diagnostic(ErrorCode.ERR_MissingDeconstruct, "new C()").WithArguments("C", "2").WithLocation(9, 34));
+        }
+
+        [Fact]
+        public void Deconstruction_TooManyDeconstructArguments()
+        {
+            var source =
+@"class C
+{
+    internal void Deconstruct(out object x, out object y) => throw null;
+}
+class Program
+{
+    static void F()
+    {
+        (object? x, object? y, object? z) = new C();
+        x.ToString();
+        y.ToString();
+        z.ToString();
+    }
+}";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (9,45): error CS1501: No overload for method 'Deconstruct' takes 3 arguments
+                //         (object? x, object? y, object? z) = new C();
+                Diagnostic(ErrorCode.ERR_BadArgCount, "new C()").WithArguments("Deconstruct", "3").WithLocation(9, 45),
+                // (9,45): error CS8129: No suitable 'Deconstruct' instance or extension method was found for type 'C', with 3 out parameters and a void return type.
+                //         (object? x, object? y, object? z) = new C();
+                Diagnostic(ErrorCode.ERR_MissingDeconstruct, "new C()").WithArguments("C", "3").WithLocation(9, 45));
+        }
+
+        // PROTOTYPE: Test:
```

Comment:
>PROTOTYPE [](start = 11, length = 9)

Do you want to move those to issues?

Reply:
It should be handled by helpers (`VisitArgumentConversion` and `Track...Assignment`), but it may be good to add tests with nested or user-defined conversions:
```C#
(List<string> x, _) = (listOfNullableString, ...);
```
```C#
(B b, _) = (someA, ...); // where A converts to B?
```
(same can be done with `Deconstruct` instead of tuple literals)

---
In reply to: [252785986](https://github.com/dotnet/roslyn/pull/32994#discussion_r252785986) [](ancestors = 252785986,252775795)

Reply:
Related to evaluation order, it may be good to test side-effecting RHS parts. 
`(x, y) = (x = null, ...);` and such

---
In reply to: [252796630](https://github.com/dotnet/roslyn/pull/32994#discussion_r252796630) [](ancestors = 252796630,252785986,252775795)

Reply:
Logged:
[https://github.com/dotnet/roslyn/issues/33017](https://github.com/dotnet/roslyn/issues/33017)
[https://github.com/dotnet/roslyn/issues/33019](https://github.com/dotnet/roslyn/issues/33019)
[https://github.com/dotnet/roslyn/issues/33021](https://github.com/dotnet/roslyn/issues/33021)
[https://github.com/dotnet/roslyn/issues/33022](https://github.com/dotnet/roslyn/issues/33022)

Regarding evaluation order, the existing `Deconstruction_EvaluationOrder_*` tests are testing side effects by dereferencing the same nullable values in multiple locations, with the warning location determined by the evaluation order.

---
In reply to: [252802108](https://github.com/dotnet/roslyn/pull/32994#discussion_r252802108) [](ancestors = 252802108,252796630,252785986,252775795)